packages/datadog-instrumentations/src/express.js

```diff
@@ -57,8 +60,125 @@ function wrapResponseRender (render) {
   }
 }
 
+function wrapAppAll (all) {
+  return function wrappedAll (path) {
+    if (routeAddedChannel.hasSubscribers) {
+      const paths = Array.isArray(path) ? path : [path]
+
+      for (const p of paths) {
+        routeAddedChannel.publish({
+          method: '*',
+          path: p instanceof RegExp ? p.toString() : p
+        })
+      }
+    }
+
+    return all.apply(this, arguments)
+  }
+}
+
+// Wrap app.route() to instrument Route object
+function wrapAppRoute (route) {
+  return function wrappedRoute (path) {
+    const routeObj = route.apply(this, arguments)
+
+    if (routeAddedChannel.hasSubscribers && typeof path === 'string') {
+      // Wrap the .all() first
+      if (typeof routeObj.all === 'function') {
+        shimmer.wrap(routeObj, 'all', (original) => function wrappedRouteAll () {
+          routeAddedChannel.publish({
+            method: '*',
+            path
+          })
+
+          return original.apply(this, arguments)
+        })
+      }
+
+      // Wrap each HTTP method
+      METHODS.forEach(method => {
+        if (typeof routeObj[method] === 'function') {
+          shimmer.wrap(routeObj, method, (original) => function wrapMethod () {
+            routeAddedChannel.publish({
+              method,
+              path
+            })
+
+            return original.apply(this, arguments)
+          })
+        }
+      })
+    }
+
+    return routeObj
+  }
+}
+
+function joinPath (base, path) {
+  if (!base || base === '/') return path || '/'
+  if (!path || path === '/') return base
+  return base + path
+}
+
+function wrapAppUse (use) {
+  return function wrappedUse () {
+    if (arguments.length < 2) {
+      return use.apply(this, arguments)
+    }
+
+    // Check if second argument has a stack (likely a router)
+    const mountPath = arguments[0]
+    const router = arguments[1]
+
+    if (typeof mountPath === 'string' && router?.stack?.length && routeAddedChannel.hasSubscribers) {
+      collectRoutesFromRouter(router, mountPath)
+    }
+
+    return use.apply(this, arguments)
+  }
+}
+
+function collectRoutesFromRouter (router, prefix) {
+  if (!router?.stack?.length) return
+
+  router.stack.forEach(layer => {
+    if (layer.route) {
+      // This layer has a direct route
+      const route = layer.route
+      const fullPath = joinPath(prefix, route.path)
+
+      Object.keys(route.methods).forEach(method => {
```

Comment:
I think this could be done in just one step with a condition only on the method which would result in simpler code. Same for the similar code above.